src/zenml/integrations/gcp/orchestrators/vertex_orchestrator.py

```diff
@@ -232,16 +245,32 @@ def _configure_container_resources(
         if memory_limit is not None:
             container_op = container_op.set_memory_limit(memory_limit)
 
-        if self.config.node_selector_constraint is not None:
-            container_op = container_op.add_node_selector_constraint(
-                label_name=self.config.node_selector_constraint[0],
-                value=self.config.node_selector_constraint[1],
-            )
-
         gpu_limit = resource_settings.gpu_count or self.config.gpu_limit
         if gpu_limit is not None:
             container_op = container_op.set_gpu_limit(gpu_limit)
 
+        if self.config.node_selector_constraint is not None:
+            constraint_label = self.config.node_selector_constraint[0]
+            value = self.config.node_selector_constraint[1]
+            if (
+                (
+                    constraint_label
+                    == GKE_ACCELERATOR_NODE_SELECTOR_CONSTRAINT_LABEL
+                    and gpu_limit is None
+                )
+                or (
+                    constraint_label
+                    == GKE_ACCELERATOR_NODE_SELECTOR_CONSTRAINT_LABEL
+                    and gpu_limit is not None
+                    and gpu_limit > 0
+                )
+                or constraint_label
+                != GKE_ACCELERATOR_NODE_SELECTOR_CONSTRAINT_LABEL
+            ):
```

Comment:
Could be done a bit shorter:

```suggestion
            if not (
                constraint_label
                == GKE_ACCELERATOR_NODE_SELECTOR_CONSTRAINT_LABEL
                and gpu_limit == 0
            ):
```